README.rst
==================
 setuptools-bower
==================

.. image:: https://travis-ci.org/inveniosoftware/setuptools-bower.png?branch=master
    :target: https://travis-ci.org/inveniosoftware/setuptools-bower
.. image:: https://coveralls.io/repos/inveniosoftware/setuptools-bower/badge.png?branch=master
    :target: https://coveralls.io/r/inveniosoftware/setuptools-bower
.. image:: https://pypip.in/v/setuptools-bower/badge.png
   :target: https://pypi.python.org/pypi/setuptools-bower/
.. image:: https://pypip.in/d/setuptools-bower/badge.png
   :target: https://pypi.python.org/pypi/setuptools-bower/

About
=====
setuptools-bower provides setuptools commands for integrating bower.

Usage
=====
setuptools-bower is on PyPI so all you need is: ::

    pip install setuptools-bower

Then from within your Python package, you can now run::

    python setup.py install_bower

This will run ```bower install``` in the directory and is thus expecting to find a ```bower.json``` and respect any configuration in a ```.bowerrc``` file.

In your ```setup.cfg``` you may setup aliases to combine bower installation with your normal install::

  [aliases]
  install_all = install_bower install

Documentation
=============
Documentation is readable at http://setuptools-bower.readthedocs.org or can be build using Sphinx: ::

    git submodule init
    git submodule update
    pip install Sphinx
    python setup.py build_sphinx

Testing
=======
Running the test suite is as simple as: ::

    python setup.py test

or, to also show test coverage: ::

    ./run-coverage.sh

License
=======
Copyright (C) 2013, 2014 CERN.

setuptools-bower is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

setuptools-bower is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with setuptools-bower; if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

In applying this licence, CERN does not waive the privileges and immunities granted to it by virtue of its status as an Intergovernmental Organization or submit itself to any jurisdiction.


CHANGES
Changelog
=========

Here you can see the full list of changes between each setuptools-bower
release.

Version 0.2.0 (released 2014-07-09)
-----------------------------------
- Adds support for Python-3.4.
- Fixes grunt build command.
- Improves documentation.

Version 0.1
-----------
- Initial public release


tests/__init__.py
# -*- coding: utf-8 -*-
##
## This file is part of setuptools-bower
## Copyright (C) 2013 CERN.
##
## setuptools-bower is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License as
## published by the Free Software Foundation; either version 2 of the
## License, or (at your option) any later version.
##
## setuptools-bower is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with setuptools-bower; if not, write to the Free Software Foundation,
## Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
##
## In applying this licence, CERN does not waive the privileges and immunities
## granted to it by virtue of its status as an Intergovernmental Organization
## or submit itself to any jurisdiction.


tests/test_install_bower.py
# -*- coding: utf-8 -*-
##
## This file is part of setuptools-bower
## Copyright (C) 2013 CERN.
##
## setuptools-bower is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License as
## published by the Free Software Foundation; either version 2 of the
## License, or (at your option) any later version.
##
## setuptools-bower is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with setuptools-bower; if not, write to the Free Software Foundation,
## Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
##
## In applying this licence, CERN does not waive the privileges and immunities
## granted to it by virtue of its status as an Intergovernmental Organization
## or submit itself to any jurisdiction.

import os
import shutil
import tempfile
from unittest import TestCase

from setuptools.dist import Distribution
from setuptools_bower.setuptools_command import BowerBuildCommand, \
    NPMBuildCommand, GruntBuildCommand


SETUP_ATTRS = {
    'name': 'bower_install_test',
    'version': '0.0',
}


SETUP_PY = """\
from setuptools import setup

setup(**%r)
""" % SETUP_ATTRS


BOWER_JSON = """
{
  "name": "bower_install_test",
  "version": "0.0.0",
  "dependencies": {
    "jquery": "~1.10.2"
  }
}
"""

BOWERRC = """
{
    "directory": "bower_components"
}
"""

PACKAGES_JSON = """
{
  "name": "npm_install_test",
  "version": "0.0.0",
  "dependencies": {
    "grunt": "*",
    "grunt-cli": "*"
  }
}
"""

GRUNTFILE_JS = """
module.exports = function(grunt) {
    grunt.registerTask('build', 'Setup test', function () {
       if (grunt.option('path') === undefined) {
            return true
       } else {
            return grunt.option('path') == 'test'
       }
    });
}
"""


class TestBowerBuildCommand(TestCase):
    def setUp(self):
        self.temp_dir = tempfile.mkdtemp()

        for filename, content in [('setup.py', SETUP_PY),
                                  ('bower.json', BOWER_JSON),
                                  ('.bowerrc', BOWERRC),
                                  ('package.json', PACKAGES_JSON),
                                  ('Gruntfile.js', GRUNTFILE_JS),
                                  ]:
            f = open(os.path.join(self.temp_dir, filename), 'w')
            f.write(content)
            f.close()

        self.old_cwd = os.getcwd()
        os.chdir(self.temp_dir)

    def tearDown(self):
        os.chdir(self.old_cwd)
        shutil.rmtree(self.temp_dir)

    def test_bower_install(self):
        """Test bower install command."""
        assert not os.path.exists(
            os.path.join(self.temp_dir, 'bower_components')
        )

        assert not os.path.exists(
            os.path.join(self.temp_dir, 'node_modules')
        )

        dist = Distribution(SETUP_ATTRS)
        dist.script_name = 'setup.py'

        cmd = NPMBuildCommand(dist)
        cmd.ensure_finalized()
        cmd.run()

        assert os.path.exists(os.path.join(self.temp_dir, 'node_modules'))

        cmd = BowerBuildCommand(dist)
        cmd.ensure_finalized()
        cmd.run()

        assert os.path.exists(os.path.join(self.temp_dir, 'bower_components'))

        cmd.force_latest = True
        cmd.run()

        cmd.production = True
        cmd.run()

        cmd = GruntBuildCommand(dist)
        cmd.ensure_finalized()
        cmd.run()

        cmd.path = 'test'
        cmd.run()


run-coverage.sh
# -*- coding: utf-8 -*-
##
## This file is part of setuptools-bower
## Copyright (C) 2013 CERN.
##
## setuptools-bower is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License as
## published by the Free Software Foundation; either version 2 of the
## License, or (at your option) any later version.
##
## setuptools-bower is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with setuptools-bower; if not, write to the Free Software Foundation,
## Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
##
## In applying this licence, CERN does not waive the privileges and immunities
## granted to it by virtue of its status as an Intergovernmental Organization
## or submit itself to any jurisdiction.
coverage run setup.py test
coverage report -m


setuptools_bower/version.py
# -*- coding: utf-8 -*-
##
## This file is part of setuptools-bower
## Copyright (C) 2013, 2014 CERN.
##
## setuptools-bower is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License as
## published by the Free Software Foundation; either version 2 of the
## License, or (at your option) any later version.
##
## setuptools-bower is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with setuptools-bower; if not, write to the Free Software Foundation,
## Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
##
## In applying this licence, CERN does not waive the privileges and immunities
## granted to it by virtue of its status as an Intergovernmental Organization
## or submit itself to any jurisdiction.

"""
Version information for setuptools-bower.

This file is imported by ``setuptools_bower.__init__``, and parsed by
``setup.py`` as well as ``docs/conf.py``.
"""

# Do not change the format of this next line. Doing so risks breaking
# setup.py and docs/conf.py

__version__ = "0.2.0"


setuptools_bower/__init__.py
# -*- coding: utf-8 -*-
##
## This file is part of setuptools-bower
## Copyright (C) 2013 CERN.
##
## setuptools-bower is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License as
## published by the Free Software Foundation; either version 2 of the
## License, or (at your option) any later version.
##
## setuptools-bower is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with setuptools-bower; if not, write to the Free Software Foundation,
## Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
##
## In applying this licence, CERN does not waive the privileges and immunities
## granted to it by virtue of its status as an Intergovernmental Organization
## or submit itself to any jurisdiction.

setuptools_bower/setuptools_command.py
# -*- coding: utf-8 -*-
##
## This file is part of setuptools-bower
## Copyright (C) 2013 CERN.
##
## setuptools-bower is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License as
## published by the Free Software Foundation; either version 2 of the
## License, or (at your option) any later version.
##
## setuptools-bower is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with setuptools-bower; if not, write to the Free Software Foundation,
## Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
##
## In applying this licence, CERN does not waive the privileges and immunities
## granted to it by virtue of its status as an Intergovernmental Organization
## or submit itself to any jurisdiction.

"""Custom setuptools commands implementation."""

import setuptools


class BowerBuildCommand(setuptools.Command):

    """Setuptools command for running ``bower <command>``."""

    description = "run bower commands."

    user_options = [
        ('bower-command=', 'c',
         'Bower command to run. Defaults to \'install\'.'),
        ('force-latest', 'F', 'Force latest version on conflict.'),
        ('production', 'p', 'Do not install project devDependencies.'),
    ]

    boolean_options = ['production', 'force-latest']

    def initialize_options(self):
        """Set default values for options."""
        self.force_latest = False
        self.production = False
        self.bower_command = 'install'

    def finalize_options(self):
        """Finalize options."""
        pass

    def run(self):
        """Execute ``bower`` command."""
        cmd = ['bower', self.bower_command]
        if self.force_latest:
            cmd.append('-F')
        if self.production:
            cmd.append('-p')
        self.spawn(cmd)


class GruntBuildCommand(setuptools.Command):

    """Setuptools command for running ``grunt <command>``."""

    description = "run grunt commands."

    user_options = [
        ('grunt-command=', 'c',
         'Grunt command to run. Defaults to \'build\'.'),
        #('force-latest', 'F', 'Force latest version on conflict.'),
        ('path=', 'p', 'Path.'),
    ]

    def initialize_options(self):
        """Set default values for options."""
        self.path = None
        self.grunt_command = 'build'

    def finalize_options(self):
        """Finalize options."""
        pass

    def run(self):
        """Execute ``grunt`` command."""
        cmd = ['grunt']
        if self.grunt_command:
            cmd.append(self.grunt_command)
        if self.path:
            cmd.append('--path=%s' % self.path)
        self.spawn(cmd)


class NPMBuildCommand(setuptools.Command):

    """Setuptools command for running ``npm <command>``."""

    description = "run NPM commands."

    user_options = [
        ('npm-command=', 'c',
         'NPM command to run. Defaults to \'install\'.'),
    ]

    def initialize_options(self):
        """Set default values for options."""
        self.npm_command = 'install'

    def finalize_options(self):
        """Finalize options."""
        pass

    def run(self):
        """Execute ``npm`` command."""
        cmd = ['npm', self.npm_command]
        self.spawn(cmd)
